src/scipp/format/formatter.py

```diff
@@ -1,40 +1,130 @@
-def format_variable(data, spec):
-    """
-    String formats the Variable according to the provided specification.
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+# @author Gregory Tucker, Jan-Lukas Wynen
+
+from typing import Any, List
+
+import numpy as np
+
+from ..core.cpp_classes import DType, Unit, Variable
+from ._parse import FormatSpec, FormatType, parse
+
+
+def format_variable(self, format_spec: str) -> str:
+    """String formats the Variable according to the provided specification.
 
     Parameters
     ----------
-    data
-        A scalar or array-like scipp Variable object
-    spec
-        Format specification; only 'c' for Compact error-reporting supported at present
+    format_spec:
+        Format specification;
+        only 'c' for Compact error-reporting supported at present.
 
     Returns
     -------
-    The formatted string
+    :
+        The formatted string.
     """
-    from numpy import array
 
-    from ..core.cpp_classes import Unit
-    dtype = str(data.dtype)
-    if not any([x in dtype for x in ('float', 'int')]) or spec is None or len(spec) < 1:
-        return data.__repr__()
-    compact = spec[-1] == 'c'
+    spec = parse(format_spec, Variable)
+    return _VARIABLE_FORMATTERS[spec.format_type](self, spec)
+
+
+def _format_sizes(data: Variable) -> str:
+    return '(' + ', '.join(f'{dim}: {size}' for dim, size in data.sizes.items()) + ')'
+
+
+def _format_unit(data: Variable) -> str:
+    if data.unit is None:
+        return '<no unit>'
+    return f'[{data.unit}]'
+
+
+def _format_element(elem: Any, spec: str):
+    if spec:
+        return f'{elem:{spec}}'
+    if isinstance(elem, (int, float)):
+        # Replicate behavior of C++ formatter.
+        return f'{elem:g}'
+    if isinstance(elem, str):
+        return f'"{elem}"'
+    return f'{elem}'
+
+
+def _as_flat_array(data):
```

Comment:
Where/why do we want to format things as a flat array?

Reply:
Because this is what `str` and `repr` do.